lib/compilers/dosbox-compiler.ts

```diff
@@ -0,0 +1,175 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import path from 'path';
+
+import fs from 'fs-extra';
+
+import {BaseCompiler} from '../base-compiler';
+import * as exec from '../exec';
+import {logger} from '../logger';
+
+export class DosboxCompiler extends BaseCompiler {
+    private readonly dosbox: string;
+    private readonly root: string;
+
+    constructor(compilerInfo, env) {
+        super(compilerInfo, env);
+
+        this.dosbox = this.compilerProps(`compiler.${this.compiler.id}.dosbox`);
+        this.root = this.compilerProps(`compiler.${this.compiler.id}.root`);
+    }
+
+    protected override async writeMultipleFiles(files: any[], dirPath: string): Promise<any[]> {
+        const filesToWrite: any[] = [];
+
+        for (const file of files) {
+            if (!file.filename) throw new Error('One of more files do not have a filename');
+
+            const fullpath = this.getExtraFilepath(dirPath, file.filename);
+            const contents = file.contents.replaceAll(/\n/g, '\r\n');
+            filesToWrite.push(fs.outputFile(fullpath, contents));
+        }
+
+        return Promise.all(filesToWrite);
+    }
+
+    protected override async writeAllFiles(dirPath: string, source: string, files: any[], filters: object) {
+        if (!source) throw new Error(`File ${this.compileFilename} has no content or file is missing`);
+
+        const inputFilename = path.join(dirPath, this.compileFilename);
+        await fs.writeFile(inputFilename, source.replaceAll(/\n/g, '\r\n'));
+
+        if (files && files.length > 0) {
+            (filters as any).dontMaskFilenames = true;
+
+            await this.writeMultipleFiles(files, dirPath);
+        }
+
+        return {
+            inputFilename,
+        };
+    }
+
+    private getDosboxArgs(tempDir: string, compileArgs: string[]) {
+        const binPath = path.relative(this.root, path.dirname(this.compiler.exe));
+        const exeName = path.basename(this.compiler.exe).replace(/\.exe$/i, '');
+        return [
+            '-c',
+            `mount c ${this.root}`,
+            '-c',
+            `mount d ${tempDir}`,
+            '-c',
+            `PATH=%PATH%;C:\\${binPath}`,
+            '-c',
+            'd:',
+            '-c',
+            `${exeName} ${compileArgs.join(' ')} > STDOUT.TXT`,
+            '-c',
+            'exit',
+        ];
+    }
+
+    private getDosboxEnv() {
+        return {
+            SDL_VIDEODRIVER: 'dummy',
+        };
+    }
+
+    protected override async execCompilerCached(compiler, args, options) {
+        if (this.mtime === null) {
```

Comment:
Presumbly this is one reason why base-compiler's `mtime` had to be made avilable.

I can't see what's different between this and the base compiler? If there's a difference maybe we can extract another subfcuntion to let us specialise just that bit instead of this copy-paste? (I'm blind and can't see the difference though 😁 )